Implement native scrolling on iOS

RobertRoeb opened the pull request:
-  Supports scrolling with finger scrolling/panning, has single pixel precision 
  - Supports SetTargetWindow() used by e.g. wxDataViewCtrl
  - Draws before window is visible (I assume for off-screen drawing into buffer)
  - Added localDeviceOrigin like in GTK+, since the native scroll window moves windows up and this needs to be corrected in wxWidgets when e.g. reporting mouse event coordintes, update region and child window positions
  - Supports native scrolling physics beyond limits, then bounces back (like standard on iOS)
  - Supports wxDataViewCtrl with no changes in its code

Commit:
Implement native scrolling on iOS

  - Supports SetTargetWindow() used by e.g. wxDataViewCtrl
  - Need to enforce single pixel steps because that is what
    finger scrolling does
  - Draws before window is visible (I assume for off-screen
    drawing into buffer)
  - Added localDeviceOrigin like in GTK+, since the native
    scroll window moves windows up and this needs to be
    corrected in wxWidgets when e.g. reporting mouse event
    coordintes, update region and child window positions
  - Supports native scrolling physics beyond limits, then
    bounces back (like standard on iOS)
  - Enforce single pixel steps in wxHtmlWindow
  - Supports wxDataViewCtrl with no changes in its code

vadz reviewed (commented):
Unfortunately I'm not competent enough to review this, @csomor could you please look at this?

FWIW I'm a bit horrified by the hacks in common code, I would hope we could somehow hide this in wxOSX implementation instead of requiring `#ifdef`s in any code using scrolling rate (IIUC).

RobertRoeb commented:
> I'm a bit horrified by the hacks in common code

These "hacks" are there to enforce that when scrolling with fingers, there are only single pixel scroll steps and this is done when setting the scrollbars. Nobody's finger uses steps of 20pixels. 
There is one other #ifdef referríng to scroll physics. The current code block scrolling outside of the requested range, but on iOS, the standard is to allow overscrolling and then to bounce back. The code does this perfectly.
Unfortunately, there is one case needed to support SetTargetWindow(). I don't think it is possible get this away, but I will look at this again.

vadz commented:
> These "hacks" are there to enforce that when scrolling with fingers, there are only single pixel scroll steps and this is done when setting the scrollbars. Nobody's finger uses steps of 20pixels.

If you say so — this doesn't seem at all obvious to me because I definitely never want to scroll in steps of 1px, whichever scrolling method is used. But I mostly would hope that we could add some `GetEffectiveScrollRate()` or whatever function and use it without having to write `#ifdef`s.

> There is one other #ifdef referríng to scroll physics. The current code block scrolling outside of the requested range, but on iOS, the standard is to allow overscrolling and then to bounce back.

Again, I would hope this could be something `wxScrolled` could take into account internally...

RobertRoeb commented:
> But does wxWidgets react to finger scrolling at all? So basically the wxPanGestureEvent

New idea: on iOS, UIScrollView captures the gesture events and then translates that into a 
new (scroll) position (actually in both x and y directions at once). So maybe we need a 
wxEVT_SCROLLWIN_PAN event with GetPixelX() and GetPixelY() (and GetPosX() and 
GetPosY() which would be the pixels positions divided by the scroll steps). On iOS, the
code would generate that directly, on other platforms, we might choose to catch the 
wxPanGestureEvent and translate that into wxEVT_SCROLLWIN_PAN in a second step.
In wxScrollHelperBase::HandleOnScroll(wxScrollWinEvent& event) we could process the
wxEVT_SCROLLWIN_PAN and translate that into a two-directional scroll event with single
pixel precision and we adapt OnPrepareDC() as well as CalcScrolledPosition() and 
CalcUnscrolledPosition() to add the small single-pixel offset. I have no computer with 
a touch screen - maybe I can simulate wxPanGestureEvent somehow.

Commit:
Added wxScrollWinPanEvent

  - Implement wxScrolled reacting to a wxScrollWinPanEvent
  - Added pixel precision offset to CalcScrolledPosition(),
    CalcUnscrolledPosition() and DoPrepareDC()
  - Make iOS UIScrollView emit that event (this should then
    also work in the rare case of horizontal and combined
    scrolling)
  - Removed nearly all #idef IPHONE code from wxScrolled
    and from wxHtmlWindow
  - I don't know if we need to change ViewStart() a partial
    line could now be visible on top during finger scrolling
    aka panning

RobertRoeb commented:
> E.g. in the scroll sample the drawing part of MyCanvas is not in synch with the scrolling, while the buttons and panels are

Are you seeing that now on macOS or iOS or both?

csomor commented:
> > > E.g. in the scroll sample the drawing part of MyCanvas is not in synch with the scrolling, while the buttons and panels are
> > 
> > 
> > Are you seeing that now on macOS or iOS or both?
> 
> iOS only, vertical scrolling keeps the text in sync, while horizontal does not, I don't know whether it's related to have only the VERTICAL event being triggered, but your code setting the device origin looks fine to me, so I don't yet know where this gets lost ...

ok, found it, it's the fact that you currently only trigger y thumbtracks, its handler then updates the scrollposition members, I'll comment this, if I do it for x and y everything stays in sync

RobertRoeb commented:
> So I think we have some permanent y offset,

The toolbar should be lower, the little square is the right place, I think. That is for another PR

Commit:
Implemented wxWindow::ScrollWindow()

  - It is now possible to scroll programmatically
  - Removed all code that pretended to interact with scrollbars
  - Added fix that stops UIScrollView from moving content around

csomor commented:
`MySubFrame` gives another challenge, I think `::ScrollWindow` has to be implemented somehow, perhaps directly manipulating contentOffset, as in this sample the scrolling is directed also to the row and column header for their respective dimensions.

csomor commented:
Ok, I've tried to find a minimal change `wxWindowMac::ScrollWindow` needs at least a `Refresh` before the return, and the `wxScrollHelperBase::HandleOnPanScroll` needs to call `m_targetWindow->ScrollWindow` as well, so that the call can be intercepted. Sorry for misleading ...

Commit:
Also call wxWindow::ScrollWindow() if UIScrolView already scrolled

  - wxScrolled<> code expects a call to ScrollWindow(), and this
    call is overridden in the scroll sample to scroll the rows
    and columns manually. So the code now calls ScrollWindow()
    but since UIScrollView already scrolled, it doesn't do
    anything in this case.
    I believe that GtkScrolledWindow also does the scrolling
    itself so I am curious how I solved the double scrolling
    issue 20 years ago.

Commit:
Implemented pan gesture support in wxScrolled<>

  - Handle wxPanGestureEvent and translate it into a
    wxScrollWinPanEvent which then scrolls in both
    direction with pixel precision.
    I have simulated wxPanGestureEvents using mouse
    dragging - I have no touch screen.
  - Like mouse wheel events (I think) the gesture events
    are generated on the target window and I am not sure
    if that works. My test case does not use SetTargetWindow()

RobertRoeb commented:
As mentioned above, I just added support for pan gestures to wxScrolled<>. I don't actually have a touch screen that can do this, so I simulated this using mouse dragging. To clarify, this is for the desktop versions and should now work on all platforms which emit the wxPanGestureEvent.

csomor commented on src/osx/iphone/window.mm:
```diff
@@ -550,6 +608,81 @@ - (void) WX_valueChangedAction:(id)sender event:(UIEvent*)event
+
+void wxWidgetIPhoneImpl::ScrollWindow( int dx, int dy, const wxRect *rect )
+{
+    if (m_blockScrollWindow) return;
+
+    wxUIView* view = (wxUIView*)m_osxView;  // wxUIView derives from UIScrollView
+
+    CGPoint position = [view contentOffset];
+
+    position.x -= dx;
+    position.y -= dy;
+
+    m_blockScrollEvents = true;
+    [view setContentOffset: position];    
+    m_blockScrollEvents = false;
+
```
at the moment please add this to get proper layered drawing in `window_osx.cpp`, in `wxWindowMac::MacDoRedraw` at the end of the `if ( bgStyle == wxBG_STYLE_ERASE )` clause in line 1962
```
                    if (!UseBgCol())
                        dc.Clear();
```

Commit:
Added clearing background color in case of wxBG_STYLE_ERASE

csomor commented:
The fact that we always have layered windows on iOS lets me see another background issue, I don't want to add too much to this PR, but this one seems important to test with scroll sample on macOS and iOS and compare, so if you agree I'd propose to add one additional change, inheriting background colors only when they are set, otherwise Clear() does always work as intended . 

in `dcclient.h` the `wxWindowDCImpl` gets an overload

```
virtual void InheritAttributes(wxWindow *win) override;
```
with an implementation in `dcclient.cpp`
```
void wxWindowDCImpl::InheritAttributes(wxWindow *win)
{
    wxCHECK_RET( win, "window can't be null" );

    SetFont(win->GetFont());

    SetTextForeground(win->GetForegroundColour());
    if ( win->UseBackgroundColour() ) {
        SetTextBackground(win->GetBackgroundColour());
        SetBackground(win->GetBackgroundColour());
    }
    SetLayoutDirection(win->GetLayoutDirection());
}
```

and use this in `wxWindowDCImpl::wxWindowDCImpl( wxDC *owner, wxWindow *window )`

after the     `DoSetClippingRegion( 0 , 0 , m_width , m_height ) ;` remove the two setters and replace with

```
    InheritAttributes(window);
```

csomor commented on src/osx/iphone/window.mm:
```diff
@@ -550,6 +608,81 @@ - (void) WX_valueChangedAction:(id)sender event:(UIEvent*)event
+
+    CGPoint position = [view contentOffset];
+    CGSize size = [view contentSize];
+    if (orient == wxVERTICAL)
+    {
+        size = CGSizeMake( size.width, range*thumb );
+        position = CGPointMake( position.x, pos*thumb );
+    } else
+    if (orient == wxHORIZONTAL)
+    {
+        size = CGSizeMake( range*thumb, size.height );
+        position = CGPointMake( pos*thumb, position.y );
+    }
```
In my scroll sample, the `MyFrame` has a scroll area that is way too big. So I think - since we must adjust this to points - that we could take the `thumb` as the part representing the view's `bounds` and expand this to `range`
or `pos` 
```
    if ( thumb == 0 )
        return;

    CGPoint position = [view contentOffset];
    CGSize size = [view contentSize];
    CGSize viewSize = [view bounds].size;

    if (orient == wxVERTICAL)
    {
        CGFloat unit = viewSize.height/thumb;
        size = CGSizeMake( size.width, range*unit );
        position = CGPointMake( position.x, pos*unit );
    }
    else if (orient == wxHORIZONTAL)
    {
        CGFloat unit = viewSize.width/thumb;
        size = CGSizeMake( range*unit, size.height );
        position = CGPointMake( pos*unit, position.y );
    }
```

would this also work in your test samples ?

Commit:
Inherit background attributes from owning window

Commit:
Correctly calculate pixelsPerLine = pixelsPerUnit in scroll target window

  - Remember those values for later use in target window
  - this sets the scroll area size correctly when given by SetScrollBar()
  - re-implement SetScrollPos() and GetScrollPos() since we
    now have pixelsPerLine and can calculate the position in units
    from the position in pixels that UIScrollView gives us

csomor reviewed (approved):
@RobertRoeb thanks a lot for your efforts, I'm fine with the PR as it is now for the entire iOS part, I've tested things with scroll sample on iOS and on macOS

csomor commented on src/osx/iphone/window.mm:
```diff
@@ -194,8 +244,15 @@ + (void)initialize
+
 void wxOSX_touchEvent(UIView* self, SEL _cmd, NSSet* touches, UIEvent *event )
 {
+    // Call the superclass handler first as that is e.g. handling mouse events for UISegementedControls
+    wxOSX_touchEventHandlerPtr superimpl = 
+        (wxOSX_touchEventHandlerPtr) [[self superclass] instanceMethodForSelector:_cmd];
+    superimpl(self, _cmd, touches, event );
+
```
@RobertRoeb can you please give me a sample on where this is needed ? Thanks

Commit:
Documentation for wxScrollWinPanEvent

csomor commented:
@vadz @RobertRoeb are there open issues on this ? I'm holding back my own iOS PR because parts of it were added to this PR, in order to make testing easier, but now I'm stuck because I don't want to break merging and I see Robert has opened another PR, which also touches on parts of this PR I'd like add ...

vadz reviewed (commented):
Some trivial fixes, please see the comment below for the more important questions.

vadz commented:
Except for the trivial issues of CI checks not passing, some typos and tons of formatting issues (I've left corrections for those I noticed but it surely would be nice if I didn't have to do it...), I'm still confused by the whole `wxScrollWinPanEvent` idea: why do we need it, when it seems that it can be made to work under iOS without any changes other than those to `wxScrolled` itself? I.e. how/why would you handle it specifically instead of letting `wxScrolled` deal with it? And, if the latter is possible, why doesn't it just do it without bothering with the new event? And if we do need the new event, it would be really great to test it on at least one other platform to ensure we don't add something we'll regret later.

If it's possible to apply just the wxiOS parts, perhaps we should do this for now to unblock further progress?

csomor commented:
> Except for the trivial issues of CI checks not passing, some typos and tons of formatting issues (I've left corrections for those I noticed but it surely would be nice if I didn't have to do it...), I'm still confused by the whole `wxScrollWinPanEvent` idea: why do we need it, when it seems that it can be made to work under iOS without any changes other than those to `wxScrolled` itself? I.e. how/why would you handle it specifically instead of letting `wxScrolled` deal with it? And, if the latter is possible, why doesn't it just do it without bothering with the new event? And if we do need the new event, it would be really great to test it on at least one other platform to ensure we don't add something we'll regret later.
> 
> If it's possible to apply just the wxiOS parts, perhaps we should do this for now to unblock further progress?

Thanks for looking at this Vadim., I understand. 

There is IMHO a need for a bidirectional scroll event, I've mentioned this some years ago, then we've discussed that we at the moment I should be emitting two events, one for the x and one for the y axis. 

That could be the workaround for iOS at the moment as well, if you want to move this out to a separate PR.  But I really think we should have this on the longer term. And the default fall through IMHO should be to send two separate scroll events if the `wxScrollWinPanEvent` has not been handled. That way we don't force adoption.

As for using another event, I really don't know. We have the `wxPanGestureEvent` but that one only indicates deltas, so it is close to the mouse wheel events, while `wxScrollWinPanEvent` indicates absolute positions, so it is close to the scroll events....

RobertRoeb commented:
> when it seems that it can be made to work under iOS without any changes other than those to `wxScrolled` itself?

Without a bi-directional scroll event, the code obviously didn't work in both directions. I initially simplified it to only scroll vertically because that is what is usually used on smartphones.

RobertRoeb commented:
> And if we do need the new event, it would be really great to test it on at least one other platform to ensure
> we don't add something we'll regret later.

I did test it simulating wxPanGestureEvents by catching mouse drag events and emitting wxPanGestureEvents then and that works as it should with smooth pixel precision scrolling on wxMac. 
If something doesn't work, we can just block the code in wxScrolled<> that catches wxPanGestureEvents - but I think that wxScrolled<> absolutely should manage finger scrolling on touch screens by default. With the new wxScrollWinPanEvent I have separeted the new behaviour into a distinct class and function. That is as clean as it can get and there is almost not #ifdef left.

For the record, UIScrollVIew catches the systems pan gesture and translates it into native scrolling. So in wxScrolled<> we do
1) Catch wxPanGestureEvent
2) Translate to wxScrollWinPanEvent
3) Call ScrollWindow
on iOS all of that is done by the native scroll view and we just update the widget to reflect the change. I am very pleased that we have been able to wrap these different behaviours into a single API that only extended the old API. Without this, we can just remove iOS support. There are no scrollbars to click on.

There is still one funny place where the old (units based) and new (pixel based) logic shows the difference. In the scroll sub example with the row and col like for a spreadsheet, the row and col scroll in units (steps of 10 or 20px) when your finger scrolls in single pixels, because the code queries GetViewStart() which uses units, not pixels. We should add a GetViewStartInPixels() GetViewStartPrecise() that adjusts for the offset pixels. That is what GetPos() and GetPosPrecise() are supposed to mean.

RobertRoeb committed:
Update src/osx/iphone/window.mm

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update src/osx/iphone/window.mm

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update src/osx/iphone/window.mm

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update src/osx/iphone/window.mm

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update include/wx/scrolwin.h

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update include/wx/osx/iphone/private.h

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update include/wx/osx/iphone/private.h

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update include/wx/osx/iphone/private.h

Co-authored-by: VZ <vz-github@zeitlins.org>

csomor commented:
> Independently of this, our macOS runner seems to be dead again :-(

oh dear, I cannot connect either, I'll check ....

csomor commented:
Perhaps another way to look at the events involved is the level of abstraction. 

`wxPanGestureEvent` is rather low level, like a mouse wheel event. Our wheel events only have one direction, i.e. if I'm using it to move a selected object I need two mouse wheel events - manhattan wise - first one then the other coordinate - ie two events, two redraws etc. While I'm having one action be the user, this gets translated into two separated events, or I'm putting them into one 2D `wxPanGestureEvent` which combines these two into one, one event, one redraw.

Now on the higher level we have scrolling, dragging scrollthumbs via mouse events in a scrollbar this mouse event gets translated into a scroll event. One direction only. If I'm having a pan gesture on a scrollable surface this means a 2-D scroll event which we don't yet have, we'd have to emulate it using two scroll events, one horizontal, one vertical, again two events, two redraws like above, or we get a 2D`wxScrollWinPanEvent` combining this.

vadz commented:
Sorry, I'm completely overwhelmed with many different things right now, so maybe I'm just not spending time on this to understand it properly, but I still think that this PR is trying to do too many different and not necessarily related things at once and that it would be better to separate them. First thing is native scrolling under iOS which, AFAIU, can be implemented without implementing any new events at wx level. Second thing is "smooth" scrolling (i.e. scrolling by pixels instead of scroll units), which can or not be desirable on the other systems too (and might be achieved there by using thumb position to determine scroll position without rounding?) but should IMO be subject of its own PR and might need to be explicitly enabled, unless we're completely sure that this is not going to break the existing code which relies on always starting a new line at physical (0, 0) position. And third thing is 2D scrolling, i.e. scrolling at once in both directions.

If we could extract and apply just the first part, it would at the very least give me some breathing space instead of being under pressure to merge this urgently because this is blocking other changes. Can we do this please?

csomor commented:
> If we could extract and apply just the first part, it would at the very least give me some breathing space instead of being under pressure to merge this urgently because this is blocking other changes. Can we do this please?

I'm fine with this if Robert sees a subset, that would work correctly, need not be perfect for the first step. If this is not possible - how granular can a PR be merged ? at the file level, or can individual chunks be picked from a file ?

csomor commented on src/osx/carbon/dcclient.cpp:
```diff
@@ -96,6 +104,20 @@ wxWindowDCImpl::~wxWindowDCImpl()
+    SetFont(win->GetFont());
+
+    SetTextForeground(win->GetForegroundColour());
+    if ( win->UseBackgroundColour() ) {
```
@vadz alternatively we might have a getter on `wxWindow` that indicates that this OS and this widget supports transparent backgrounds, and then - on such an instance - not generating a color if `GetBackgroundColour` is asked on a window that doesn't have it set, or just not calling the method, ie moving this code to base either to `GetBackgroundColour` or to `InheritAttributes`, and extend it with something like

```
if ( !win->SupportsTransparentBackground() && win->UseBackgroundColor() )
{
        SetTextBackground(win->GetBackgroundColour());
        SetBackground(win->GetBackgroundColour());
}
```

Or I'd have to modifiy the default attributes in the base class when running on OSX to always have wxTransparentColour by default as background, button face is not a good choice for these systems.

Commit:
Removed wxScrollWinPanEvent

  - Support iOS scrolling through two SCROLLWIN_THUMBTRACK events
  - Removed support for wxPanGestureEvent from scrolling
  - re-added #ifdef __WXOSX_IPHONE__ to allow overscrolling
  - Added GetPixelOffset() to wxScrollWinEvent for single pixel
    precision when UIScrollWin scrolls by single pixels

vadz committed:
Minor style fixes

Apply suggestions from the previous reviews.

vadz reviewed (commented):
Thanks, there is still the issue of background colour inheriting remaining, but I guess we'll deal with (or not...) later.

I'll address the remaining issues (i.e. remove the leftovers) later if you don't beat me to it.

RobertRoeb committed:
Update src/osx/iphone/window.mm

Co-authored-by: VZ <vz-github@zeitlins.org>

RobertRoeb committed:
Update include/wx/osx/core/private.h

Co-authored-by: VZ <vz-github@zeitlins.org>

Commit:
Removed leftovers from wxScrollWinPanEvent and updated docs

  - Documented wxScrollHelper::GetViewStartPixels() and wxScrollEvent::GetPixelOffset()

Commit:
Corrected GetViewStartPixels() and Calculation of ScrollWindow

  - Corrected copy/paste error in GetViewStartPixels()
  - Now correctly calculate ScrollWindow() delta if using pixel
    precision
  - This allows the columns and rows in the scroll sub sample
    to also move with pixel precision when using GetViewStartPixels()
    in their OnPaint() methods

Commit:
Remove white spaceRemove white spacess

Commit:
Remove white spaces

Commit:
Remove white space

vadz commented:
I'll apply and merge it now.
